### PR TITLE
Fix lookup functions `MATCH`, `VLOOKUP` and `HLOOKUP`

### LINKED ISSUE #1063: MATCH using -1 as type
```
MATCH(Searchcriterion; Lookuparray; Type)
```
The match function specifies a type as the 3rd argument. If you use "1", HF works correctly to match the lower bound. However, if you use "-1" it should return the upper bound. This is the behavior of Sheets and Excel. Hyperformula throws an error.

### MATCH working correctly
```
test("Match with negative match", () => {
  const engine = HyperFormula.buildFromArray(
    [["=MATCH(203, A2:A5, 1)"], ["100"], ["200"], ["300"], ["400"], ["500"]],
    {
      licenseKey: "gpl-v3",
    }
  );

  expect(engine.getCellValue({ sheet: 0, row: 0, col: 0 })).toEqual(2);
});
```
### MATCH throwing error

```
Error: Value of the formula cell is not computed.
```

```
test("Match with negative match", () => {
  const engine = HyperFormula.buildFromArray(
    [["=MATCH(203, A2:A5, -1)"], ["500"], ["400"], ["300"], ["200"], ["100"]],
    {
      licenseKey: "gpl-v3",
    }
  );

  expect(engine.getCellValue({ sheet: 0, row: 0, col: 0 })).toEqual(3);
});
```

I should note, this assumes that the data is ordered correctly. We should probably update the documentation to mention that requirement. When the data is ordered incorrectly, Google sheets will return the largest value, whereas hyperformula returns the same value whether its -1 or 1. We may want to update the runtime differences docs to reflect that difference if it is intended. See below for that example. You can see Google sheets explain the data ordering assumption here: https://support.google.com/docs/answer/3093378

```
test("Match with negative match", () => {
  const engine = HyperFormula.buildFromArray(
    [["=MATCH(203, A2:A5, -1)"], ["100"], ["200"], ["300"], ["400"], ["500"]],
    {
      licenseKey: "gpl-v3",
    }
  );

  expect(engine.getCellValue({ sheet: 0, row: 0, col: 0 })).toEqual(5); // Google sheets returns 5 in this scenario, hyperformula returns 2. 
});
```

### LINKED ISSUE #1072: `VLOOKUP` and `HLOOKUP` return the first match instead of the last one when RangeLookup = TRUE
### Description

From [ODFF](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html#__RefHeading__1018422_715980110111111111111111111111111111):

> If RangeLookup is omitted or TRUE or not 0, the first row of DataSource is assumed to be sorted in ascending order, with smaller numbers before larger ones [...] The lookup will try to match an entry of value Lookup. If none is found the largest entry less than Lookup is taken as a match. From a sequence of identical values ≤ Lookup the last entry is taken.

### Steps to reproduce

These tests fail:

```js
    it('VLOOKUP should return the last matching value if RangeLookup = TRUE', () => {
      const engine =HyperFormula.buildFromArray([
        ['1', 'a'],
        ['2', 'b'],
        ['2', 'c'],
        ['2', 'd'],
        ['2', 'e'],
        ['=VLOOKUP(2, A1:B5, 2, TRUE())'],
      ])

      expect(engine.getCellValue(adr('A6'))).toEqual('e')
    })

    it('HLOOKUP should return the last matching value if RangeLookup = TRUE', () => {
      const engine = HyperFormula.buildFromArray([
        ['1', '2', '2', '2', '5'],
        ['a', 'b', 'c', 'd', 'e'],
        ['=HLOOKUP(2, A1:E2, 2, TRUE())'],
      ])

      expect(engine.getCellValue(adr('A3'))).toEqual('d')
    })
```

### Links
Related: #1063

### PR DESCRIPTION
**Pull request by sequba**
### Description of changes

- rewrite the binary search algorithm to be able to handle the ranges ordered descending
- rewrite the `SearchStrategy.find()` method in all the child classes
- refactor some other methods in `ColumnIndex` class
- adjust the implementation of function MATCH
- unit tests:
  - add tests for MATCH with `MatchType = -1` (#1063)
  - add tests for #1072
  - add more tests for MATCH, VLOOKUP and HLOOKUP for different edge cases and refactor the test suites
  - improve test utils `expectArrayWithSameContent` and `resetSpy`

### How did you test your changes?

- unit tests passed (including lots of new unit tests for the changed features)

### Types of changes

- [ ] Breaking change (a fix or a feature because of which an existing functionality doesn't work as expected anymore)
- [ ] New feature or improvement (a non-breaking change that adds functionality)
- [x] Bug fix (a non-breaking change that fixes an issue)
- [ ] Additional language file, or a change to an existing language file (translations)
- [ ] Change to the documentation

### Related issues:
- Fixes #1063
- Fixes #1072

### Checklist:

- [x] My change is compliant with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard.
- [x] My change is compatible with Microsoft Excel.
- [x] My change is compatible with Google Sheets.
- [x] My code follows the code style of this project.
- [x] I described my changes in the [CHANGELOG.md](https://github.com/handsontable/hyperformula/blob/master/CHANGELOG.md) file.
- [ ] My changes require a documentation update.
- [ ] My changes require a migration guide.

### CODE REVIEW

**Review by budnix — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
+- Added a new mode to the MATCH function that searches in ranges ordered descending. [#1063](https://github.com/handsontable/hyperformula/issues/1063)
```
[the remark this answers is not in the document]
Comment (warpech) [reply CHANGELOG.md:10]: I think it is a bugfix and I think it should be presented as such in the changelog

[the remark this answers is not in the document]
Comment (sequba) [reply CHANGELOG.md:10]: Ok: https://github.com/handsontable/hyperformula/pull/1092
